worlds/rogue_legacy/__init__.py

```diff
@@ -31,215 +36,161 @@ class RLWorld(World):
+                "Enchantress - Cape",
+            ]
+            self.multiworld.local_early_items[self.player][self.random.choice(possible_items)] = 1
+
+    def create_item(self, name: str) -> RLItem:
+        item_data = item_table[name]
+        return RLItem(name, item_data.classification, item_data.code, self.player)
 
     def create_items(self):
         item_pool: List[RLItem] = []
-        total_locations = len(self.multiworld.get_unfilled_locations(self.player))
-        for name, data in item_table.items():
-            quantity = data.max_quantity
-
-            # Architect
-            if name == "Architect":
-                if self.get_setting("architect") == "disabled":
-                    continue
-                if self.get_setting("architect") == "start_unlocked":
-                    self.multiworld.push_precollected(self.create_item(name))
-                    continue
-                if self.get_setting("architect") == "early":
-                    self.multiworld.local_early_items[self.player]["Architect"] = 1
-
-            # Blacksmith and Enchantress
-            if name == "Blacksmith" or name == "Enchantress":
-                if self.get_setting("vendors") == "start_unlocked":
-                    self.multiworld.push_precollected(self.create_item(name))
-                    continue
-                if self.get_setting("vendors") == "early":
-                    self.multiworld.local_early_items[self.player]["Blacksmith"] = 1
-                    self.multiworld.local_early_items[self.player]["Enchantress"] = 1
-
-            # Haggling
-            if name == "Haggling" and self.get_setting("disable_charon"):
-                continue
+        location_count = len(self.multiworld.get_unfilled_locations(self.player))
+        self.fountain_piece_requirement = 0
+
+        # Create each item we can create.
+        for item_data in item_table.values():
+            item_pool += [
+                self.create_item(item_data.name)
+                for _ in range(item_data.creation_quantity(self.multiworld, self.player))
+            ]
+
+        # Handle fountain piece creation.
+        if self.get_setting("fountain_door_requirement") != "bosses":
+            maximum_fountain_pieces: int = self.get_setting("fountain_pieces_available").value
+            remaining_item_slots = location_count - len(item_pool)
+            fountain_pieces = [
+                self.create_item("Piece of the Fountain")
+                for _ in range(min(maximum_fountain_pieces, remaining_item_slots))
+            ]
+
+            # Calculate requirement and add to item pool.
+            percentage: int = self.get_setting("fountain_pieces_percentage").value
+            self.fountain_piece_requirement = round(max(len(fountain_pieces) * (percentage / 100), 1))
+            item_pool += fountain_pieces
+
+            # Log a warning if not enough locations were able to create requested pieces.
+            warning(
```

Comment:
This warning is not in an `if` block and appears regardless of whether fountain pieces have been reduced